### ISSUE TITLE
Performance Concerns with "Radon" and "IRadon" Operations

### ISSUE THREAD

**Issue by Guaishou74851** [deepinv/deepinv#48]
**Background:**

I am utilizing the "Radon" and "IRadon" classes from [tomography.py](https://github.com/deepinv/deepinv/blob/main/deepinv/physics/tomography.py) to conduct sparse-view CT reconstruction experiments. Here's my code snippet for initializing these classes:

```python
view_number = 50
theta = np.linspace(0, 180, view_number, endpoint=False)
A = Radon(256, theta).to(device)
A_dagger = IRadon(256, theta).to(device)
```

I set the `device` variable to `torch.device('cuda')`. However, I have encountered two significant issues that require attention.

**1. Warning during execution:**
Upon running the code, I receive the following warning message:
```
"/opt/conda/lib/python3.8/site-packages/torch/functional.py:568: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  ../aten/src/ATen/native/TensorShape.cpp:2228.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]"
```
This warning may be attributed to the code in line 219 of [tomography.py](https://github.com/deepinv/deepinv/blob/main/deepinv/physics/tomography.py) (`return torch.meshgrid(unitrange, unitrange)`).

**2. CPU reliance during forward operations:**
The forward operations of the "Radon" and "IRadon" classes seem to heavily rely on the CPU, rather than making optimal use of available GPU resources. In my case, I am training a deep proximal gradient descent (PGD) algorithm-unrolled network with ten unrolled iterations. Each iteration involves the operation "x = x - rho * A_dagger(A(x) - y)". Here, "x" represents a batch of image tensors with a shape of (1, 1, 256, 256) from the CT100 dataset, and "rho" is a scalar value (e.g., 0.01).

**Observations:**
When training the network using a single Python program on a Linux computer equipped with four NVIDIA GTX 4090 GPUs, I achieved a training speed of approximately 2.5 iterations per second, with the utilization of single GPU at around 25%. However, when running three simultaneous instances of the training program, each utilizing a different GPU, the training speed per program dropped to roughly 0.3 iterations per second, and the GPU utilization per GPU decreased to approximately 5%.

**Request:**
May I venture to ask whether the forward operations of the "Radon" and "IRadon" (with filtering) classes heavily rely on the CPU rather than efficiently utilizing the GPU? Furthermore, I would greatly appreciate any guidance or suggestions to address this performance issue and enhance the training process. Specifically, I am interested in mitigating the problem while considering scenarios such as running **4** training instances, each performing **10** "Radon" and "IRadon" forward operations within the deep unrolled network's forward process. Additionally, if you have any recommendations to improve GPU utilization (e.g., increasing from 5% to 25% or beyond), they would be highly valued.

Thank you very much for your attention to this matter. I eagerly await your valuable insights and assistance.

**Comment by samuro95** [comment]
Thanks for reporting this issue. Could you provide your training code in order to reproduce your issue ?

**Comment by samuro95** [comment]
Thanks for reporting this issue. Could you provide your training code in order to reproduce your issue ?

**Comment by Guaishou74851** [comment]
**Description:**

**Issue 1: Warning regarding `torch.meshgrid`**

When running the provided code, the following warning is displayed:

```
/opt/conda/lib/python3.8/site-packages/torch/functional.py:568: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  ../aten/src/ATen/native/TensorShape.cpp:2228.)
```

**Issue 2: Decreased training speed and GPU utilization with multiple GPU instances**

When executing the training program with a single instance on a single 4090 GPU, the training speed is approximately 2.5 iterations per second. However, when running four program instances simultaneously on four different 4090 GPUs on the same computer, the speed decreases significantly to about 12 seconds per iteration.

**Code:**

```Python
import os
import torch
from torch import nn
import numpy as np
import deepinv
from tqdm import tqdm

os.environ['CUDA_DEVICE_ORDER'] = 'PCI_BUS_ID'
os.environ['CUDA_VISIBLE_DEVICES'] = '0'

view_number = 50
angles = np.linspace(0, 180, view_number, endpoint=False)
CT = deepinv.physics.Tomography(256, angles)

class Stage(nn.Module):
    def __init__(self):
        super().__init__()
        self.rho = nn.Parameter(torch.tensor([0.01]))
        self.body = nn.Sequential(
            nn.Conv2d(1, 32, 3, padding=1), nn.ReLU(True),
            nn.Conv2d(32, 32, 3, padding=1), nn.ReLU(True),
            nn.Conv2d(32, 32, 3, padding=1), nn.ReLU(True),
            nn.Conv2d(32, 32, 3, padding=1), nn.ReLU(True),
            nn.Conv2d(32, 32, 3, padding=1), nn.ReLU(True),
            nn.Conv2d(32, 1, 3, padding=1),
        )

    def forward(self, x):
        x = x - self.rho * CT.A_dagger(CT.A(x) - y)
        x = x + self.body(x)
        return x

class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.body = nn.Sequential(*[Stage() for _ in range(10)])

    def forward(self, x):
        return self.body(x)

model = Net().cuda()
optimizer = torch.optim.AdamW(model.parameters(), lr=1e-4, weight_decay=0.0)

print('start training ...')
epoch_num = 100
iter_num = 100
for i in range(epoch_num):
    loss_avg = 0.0
    for j in tqdm(range(iter_num)):
        x = torch.rand(1, 1, 256, 256, device='cuda')
        y = CT.A(x)
        loss = (model(CT.A_dagger(y)) - x).pow(2).mean()
        optimizer.zero_grad(set_to_none=True)
        loss.backward()
        optimizer.step()
        loss_avg += loss.item()
    loss_avg /= iter_num
    print('Epoch %d, average loss: %f.' % (i, loss_avg))
```

Please note that the code assumes `torch.__version__` to be '1.11.0+cu113'.

Thank you very much for your attention!

**Comment by Guaishou74851** [comment]
> Thanks for sharing your code! Not all computations were performed on GPU; we fixed it in #46 and pushed on main. After pulling from main, the following line of your code needs to be updated (angles are now torch tensors).
> 
> ```python
> angles = torch.from_numpy(np.linspace(0, 180, view_number, endpoint=False))
> ```
> 
> Running your code on 4 different GPUs (V100) simultaneously slightly decreases the training time, moving from ~2.20s/it on average for the 1st epoch to ~1.90s/it when the 4 GPUs are running.
> 
> Does this solve your issue?

I have updated the code and found that **(1)** the warning has been eliminated, and **(2)** the training speed has significantly improved.

Thank you so much for your invaluable assistance!

**Comment by Guaishou74851** [comment]
**Issue Response: Discrepancy in Output of "Radon" and "IRadon" Operations**

I have observed a potential discrepancy in the output of the "Radon" and "IRadon" operations between the old and new versions of `deepinv.physics.Tomography`. Here are the details of my observation:

Following (Dr. Chen's REI)[https://github.com/edongdongchen/REI], I am working with the [CT100](https://www.kaggle.com/datasets/kmader/siim-medical-images) dataset, specifically the last 10 test images, using the following code for compressive sampling and initialization:

```Python
...
view_number = 50
angles = np.linspace(0, 180, 50, endpoint=False)    # should be converted to torch tensor for the latest version of deepinv.physics.Tomography
CT = deepinv.physics.Tomography(256, angles)
...
x = torch.tensor(image).cuda()    # image, shape is (1, 1, 256, 256)
y = CT.A(x)    # measurement, shape is (1, 1, 363, 50)
x_init = CT.A_dagger(y)    # recovered image, shape is (1, 1, 256, 256)
psnr = PSNR(x, x_init)    # calculate the PSNR between the ground truth and recovered image
...
```

With the old version of `deepinv.physics.Tomography` (prior to the updates in #46), I obtained an average PSNR of 28.60dB on the 10 CT100 test images. However, when using the latest version of `deepinv.physics.Tomography` (following the updates in #46), the average PSNR dropped to 25.75dB.

While I appreciate that the new implementations of "Radon" and "IRadon" operations are now warning-free and faster, I would like to inquire if it is expected for the outputs of these two versions to differ slightly.

Thank you very much for your great efforts, significant contribution, and attention. I await clarification on this matter.

**Comment by matthieutrs** [comment]
Thanks for raising this! We actually witnessed discrepencies between results depending on the datatypes and are working on resolving this. While this accounts for a non-negligible ~.1 dB difference, the number you mention is much higher...

May I ask how you generate `angles` in a torch format? In fact, `torch.linspace()` and `np.linspace()` differ when `endpoint=True` in the numpy case. Below is an example showing the difference:

```python
import numpy as np
import torch
import deepinv

device = "cuda"
dtype = torch.double

# Testing pure numpy input, cast to torch
view_number = 50
angles = torch.from_numpy(np.linspace(0, 180, view_number,  endpoint=False)).to(device).type(dtype)
CT = deepinv.physics.Tomography(256, angles, device=device, dtype=dtype)

torch.manual_seed(0)
x = torch.rand(1, 1, 256, 256, device=device, dtype=dtype)
y = CT.A(x)
z = CT.A_dagger(y)
psnr = deepinv.utils.cal_psnr(x, z)
print('from numpy : ', psnr)


# Testing pure torch, ignoring endpoint=False
angles = torch.linspace(0, 180, view_number, device=device, dtype=dtype)
CT = deepinv.physics.Tomography(256, angles, device=device, dtype=dtype)

torch.manual_seed(0)
x = torch.rand(1, 1, 256, 256, device=device, dtype=dtype)
y = CT.A(x)
z = CT.A_dagger(y)
psnr = deepinv.utils.cal_psnr(x, z)
print('torch.linspace() : ', psnr)


# Testing pure torch, taking endpoint=False into account
angles = torch.linspace(0, 180, view_number+1, device=device, dtype=dtype)[:-1]
CT = deepinv.physics.Tomography(256, angles, device=device, dtype=dtype)

torch.manual_seed(0)
x = torch.rand(1, 1, 256, 256, device=device, dtype=dtype)
y = CT.A(x)
z = CT.A_dagger(y)
psnr = deepinv.utils.cal_psnr(x, z)
print('torch.linspace() adjusted : ', psnr)
```

This outputs:
```
torch.from_numpy(np.linspace()) :  9.396799382665893
torch.linspace() :  8.307137213612316
torch.linspace() adjusted :  9.396799382665895
```

Notice the 1.1dB difference between the torch.linspace() and its adjusted version, equal to the torch.from_numpy(np.linspace()) one. 

Would it be possible that your drop in PSNR comes from there?

**Comment by Guaishou74851** [comment]
Thank you very much for your efforts and replying my concerns.
Actually, this answer may not exactly hit what I mean. Please refer to the attached codes: 
[test.zip](https://github.com/deepinv/deepinv/files/11925427/test.zip).

**Comment by matthieutrs** [comment]
Thanks for sharing your codes! It turns out that your PSNR discrepency comes from the default `"circle"` parameter: setting 
```python
class Radon(nn.Module):
    def __init__(
    ...
    circle=False,
    ...
```
and same in `IRadon` solves the issue on my side; I've created a PR with this is fix in #59.

Let me know if this does not solve your issue!

**Comment by Guaishou74851** [comment]
All my issues in this thread have been resolved. I would like to express my sincere appreciation to @samuro95 and @matthieutrs for their excellent support and follow-up throughout this process.
